api_app/api.py

```diff
@@ -421,6 +423,58 @@ def download_sample(request):
         )
 
 
+@api_view(["PATCH"])
+def kill_running_job(request):
+    """
+    kill running job by closing celery tasks and marking as killed
+    :param request: job_id
+    :returns: 200 if killed, 404 not found, 403 forbidden, 400 bad request
+    """
+    try:
+        data_received = request.query_params
+        logger.info(
+            f"kill_running_job received request from {str(request.user)}."
+            f"Data received {data_received}."
+        )
+        if "job_id" not in data_received:
+            return Response({"error": "821"}, status=status.HTTP_400_BAD_REQUEST)
+        # get job object
+        try:
+            job = models.Job.objects.get(id=data_received["job_id"])
+        except models.Job.DoesNotExist:
+            return Response(
+                {"detail": "job not found"}, status=status.HTTP_404_NOT_FOUND
+            )
+        # check permission
+        if not request.user.has_perm("api_app.change_job", job):
+            return Response(
+                {"detail": "You don't have permission to perform this operation."},
+                status=status.HTTP_403_FORBIDDEN,
+            )
+        # check if job running
+        if job.status != "running":
+            return Response(
+                {"detail": "job not running"},
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+        # close celery tasks
+        task_ids = cache.get(data_received["job_id"])
+        if isinstance(task_ids, list):
+            celery_app.control.revoke(task_ids)
+            cache.delete((data_received["job_id"]))
+        # set job status
+        job.status = "killed"
+        job.save()
```

Comment:
I would prefer if this is made as an utility function and kept in general.py as `kill_running_analysis`.

Reply:
This can be resolved.

Reply:
This can be resolved too.